// ng-zorro-antd/esm2015/experimental/image/image.module.js
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { BidiModule } from '@angular/cdk/bidi';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { OverlayModule } from '@angular/cdk/overlay';
import { PlatformModule } from '@angular/cdk/platform';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NzIconModule } from 'ng-zorro-antd/icon';
import { NzImageModule as ImageModule } from 'ng-zorro-antd/image';
import { NzPipesModule } from 'ng-zorro-antd/pipes';
import { NzImageViewComponent } from './image.component';
import * as ɵngcc0 from '@angular/core';
export class NzImageModule {
}
NzImageModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NzImageModule });
NzImageModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NzImageModule_Factory(t) { return new (t || NzImageModule)(); }, imports: [[
            BidiModule,
            OverlayModule,
            PortalModule,
            DragDropModule,
            CommonModule,
            NzIconModule,
            NzPipesModule,
            PlatformModule,
            ImageModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NzImageModule, { declarations: function () { return [NzImageViewComponent]; }, imports: function () { return [BidiModule,
        OverlayModule,
        PortalModule,
        DragDropModule,
        CommonModule,
        NzIconModule,
        NzPipesModule,
        PlatformModule,
        ImageModule]; }, exports: function () { return [NzImageViewComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzImageModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BidiModule,
                    OverlayModule,
                    PortalModule,
                    DragDropModule,
                    CommonModule,
                    NzIconModule,
                    NzPipesModule,
                    PlatformModule,
                    ImageModule
                ],
                exports: [NzImageViewComponent],
                declarations: [NzImageViewComponent]
            }]
    }], null, null); })();
