// ng-zorro-antd/esm2015/cdk/overflow/overflow.module.js
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { NgModule } from '@angular/core';
import { NzResizeObserverModule } from 'ng-zorro-antd/cdk/resize-observer';
import { NzOverflowContainerComponent } from './overflow-container.component';
import { NzOverflowItemDirective } from './overflow-item.directive';
import { NzOverflowRestDirective } from './overflow-rest.directive';
import { NzOverflowSuffixDirective } from './overflow-suffix.directive';
import * as ɵngcc0 from '@angular/core';
export class NzOverflowModule {
}
NzOverflowModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NzOverflowModule });
NzOverflowModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NzOverflowModule_Factory(t) { return new (t || NzOverflowModule)(); }, imports: [[NzResizeObserverModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NzOverflowModule, { declarations: function () { return [NzOverflowContainerComponent,
        NzOverflowItemDirective,
        NzOverflowRestDirective,
        NzOverflowSuffixDirective]; }, imports: function () { return [NzResizeObserverModule]; }, exports: function () { return [NzOverflowContainerComponent, NzOverflowItemDirective, NzOverflowRestDirective, NzOverflowSuffixDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzOverflowModule, [{
        type: NgModule,
        args: [{
                imports: [NzResizeObserverModule],
                declarations: [
                    NzOverflowContainerComponent,
                    NzOverflowItemDirective,
                    NzOverflowRestDirective,
                    NzOverflowSuffixDirective
                ],
                exports: [NzOverflowContainerComponent, NzOverflowItemDirective, NzOverflowRestDirective, NzOverflowSuffixDirective]
            }]
    }], null, null); })();
