// ng-zorro-antd/esm2015/cdk/resize-observer/resize-observer.module.js
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { NgModule } from '@angular/core';
import { NzResizeObserverDirective } from './resize-observer.directive';
import { NzResizeObserverFactory } from './resize-observer.service';
import * as ɵngcc0 from '@angular/core';
export class NzResizeObserverModule {
}
NzResizeObserverModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NzResizeObserverModule });
NzResizeObserverModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NzResizeObserverModule_Factory(t) { return new (t || NzResizeObserverModule)(); }, providers: [NzResizeObserverFactory] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NzResizeObserverModule, { declarations: function () { return [NzResizeObserverDirective]; }, exports: function () { return [NzResizeObserverDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzResizeObserverModule, [{
        type: NgModule,
        args: [{
                providers: [NzResizeObserverFactory],
                declarations: [NzResizeObserverDirective],
                exports: [NzResizeObserverDirective]
            }]
    }], null, null); })();
