// ngx-csv-parser/esm5/lib/ngx-csv-parser.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var NgxCsvParserComponent = /** @class */ (function () {
    function NgxCsvParserComponent() {
    }
    NgxCsvParserComponent.prototype.ngOnInit = function () {
    };
NgxCsvParserComponent.ɵfac = function NgxCsvParserComponent_Factory(t) { return new (t || NgxCsvParserComponent)(); };
NgxCsvParserComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgxCsvParserComponent, selectors: [["lib-ngx-csv-parser"]], decls: 2, vars: 0, template: function NgxCsvParserComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "p");
        ɵngcc0.ɵɵtext(1, " ngx-csv-parser works! ");
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxCsvParserComponent, [{
        type: Component,
        args: [{
                selector: 'lib-ngx-csv-parser',
                template: "\n    <p>\n      ngx-csv-parser works!\n    </p>\n  "
            }]
    }], function () { return []; }, null); })();
    return NgxCsvParserComponent;
}());
export { NgxCsvParserComponent };
