// ngx-csv-parser/esm5/lib/ngx-csv-parser.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { NgxCsvParserComponent } from './ngx-csv-parser.component';
import { NgxCsvParser } from './ngx-csv-parser.service';
import * as ɵngcc0 from '@angular/core';
var NgxCsvParserModule = /** @class */ (function () {
    function NgxCsvParserModule() {
    }
NgxCsvParserModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxCsvParserModule });
NgxCsvParserModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxCsvParserModule_Factory(t) { return new (t || NgxCsvParserModule)(); }, providers: [NgxCsvParser], imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxCsvParserModule, { declarations: function () { return [NgxCsvParserComponent]; }, exports: function () { return [NgxCsvParserComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxCsvParserModule, [{
        type: NgModule,
        args: [{
                declarations: [NgxCsvParserComponent],
                imports: [],
                providers: [NgxCsvParser],
                exports: [NgxCsvParserComponent]
            }]
    }], function () { return []; }, null); })();
    return NgxCsvParserModule;
}());
export { NgxCsvParserModule };
